### DISCUSSION TITLE
Enable Tree-View interface for O2M relations

### DISCUSSION THREAD

**Discussion by LasseRosenow** [directus/directus#7447] [Feature Requests]
I think it would be great, if the Tree-View interface would also be enabled on O2M relations.

I guess the reason why this is not done right now is, because not all items in a nested tree-view are necessarily from the same collection, so you cannot just move everything around and make each item a child of each other item.

But that could be fixed by for example checking the whole tree, when starting to drag an item, if it can have the currently being dragged item as a child. So one way would be to give all the items that cannot have the currently dragged item as a child, a dark background or make them more transparent or something like that just so you can see, where its possible to release the item and where not.

## Use-Cases:
There are many use-cases for this. 

### Menus:
For example it could improve building a navigation-menu by having a `main-navigation` single instance collection which has a O2M dependency to a recursive `navigation-item` collection. Right now you would only see the top-level items within the single instance collection and would only have a tree for all the children..
Right-now:
<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td>

![image](https://user-images.githubusercontent.com/10547444/129641392-d6dccfbb-bed4-48c9-a694-a0c06fd8499e.png)

</td>
<td>

![Unbenannt](https://user-images.githubusercontent.com/10547444/129643762-801c5c8d-497a-48b0-a458-3a8c31e2c379.png)

</td>
</tr>
</table>

### Page Content
This one would need O2A relations first, but that would enable us to write the dynamic content of a page with the help of tree structures:
![Unbenannt](https://user-images.githubusercontent.com/10547444/129643568-2a712c97-864f-450d-9d98-08b4b66e1623.png)

**Comment by benhaynes** [comment]
Thanks! Very comprehensive feature request... which I always appreciate. Have you looked into M2A relationships? It seems like those might be better suited for this multi-collection sorting.

I'm not quite sure how you would configure which collections / fields should be shown by this O2M version... and you're right that's sorting would be difficult. Might be worth leaving out sorting for an MVP.

**Reply by LasseRosenow** [reply]
Hey thank you for your feedback :)

> Have you looked into M2A relationships? It seems like those might be better suited for this multi-collection sorting.

The issue with M2A relationships is that, they would only give me a flat hierarchy and not a tree structure. Also M2A would not work well in tree structures, as `A` can have multiple `M` within the same tree, so you could have the same item multiple times within the tree. I think that could be very confusing and would also not be very necessary in a Tree-View scenario. Hence `O2A` as `O2A` would have a clear relation where each `A` only has 1 parent.


> I'm not quite sure how you would configure which collections / fields should be shown by this O2M version...

I suppose you mean the case when a collection contains multiple `O2M` relations right?
I think there is no easy answer to this. 
- One way would be to make a constraint that the tree-view interface can only be enabled on 1 `O2M` relation per collection.
- Another way would be to just ignore collections that have multiple `O2M` relations for now.
- I think the best solution would be to think of a way to visualize it somehow within the graph. Maybe draw boxes around each `O2M` list of a relation. I add a small sketch of how it could be done:

![Unbenannt](https://user-images.githubusercontent.com/10547444/129731186-26417271-8280-4271-ab2a-c602bfe0c30f.png)
So in this sketch the first white box contains all children of a `O2M` relation to collection `x`, while the second white box contains all the children of a `O2M` relation to the collection `y`
Of course this could be made to look much better, but I'm not so much of a designer myself 😅

> Might be worth leaving out sorting for an MVP.

Yes same goes for my more advanced `O2M` proposal. Maybe just ignore collections with multiple `O2M` relations for now and iterate from there :)


I hope I covered all your concerns, but if there are still things not clear enough, just let me know :)

**Reply by benhaynes** [reply]
For the M2A, remember that we can add whatever functionality we need! In addition to the `sort` we have now, there could be a configurable `parent` field, that allows nesting. We can also look into supporting _unique_ constraints either at the API-level (code) or the db-level (compound keys). Just throwing this out there as an alternate direction.

Yeah, for the O2M, not just multiple O2M, but when to stop! I could imagine all sorts of recursive issues, or rambling relationships that go deeper than you care to manage (eg, you may want to link Files, but then that keeps going to Users (who uploaded it), then to Roles... etc. I'm not 100% sold on how 80/20 the O2M option is, but if it were to be created, it would need to have an "allow-list" of what relationships to track/allow. Just my two cents here. 👍 

BTW, those mockups are great... especially if you're not a designer! :)

**Reply by rijkvanzanten** [reply]
This does mean that you'd lose the ability to have deeply-nested structures right? The reason the current version is limited to a single recursive collection is that it's, well, recursive! So you can have 

```
item
  sub-item
    sub-sub-item
```

with sorting ootb. When you'd use a "regular" o2m one, you'd get something like

```
author
  article
  article
author
  article
```

but you couldn't go any deeper, as there is no relationship for the children of "article" 🤔

**Reply by LasseRosenow** [reply]
@benhaynes 
> For the M2A, remember that we can add whatever functionality we need! In addition to the sort we have now, there could be a configurable parent field, that allows nesting.

True that could also be done, but I think that would make things more complicated and also confusing as you would then be able to have a tree, but some M2A items are only children of one parent, because it was configured somewhere .. even though in reality they have multiple parents. O2A on the other hand would just work as expected

> Yeah, for the O2M, not just multiple O2M, but when to stop!

True point. First of all I think that this would also be some edge case like how you can create circular nesting in the current tree-view implementation. I think the benefit outweighs the loss, but there are things to work around this as well of course:
- I think the tree should only display nested relations that enabled the tree-view interface themselves. 
- On top of that you could make it configurable how deep the nesting should go at the same place, where the tree-view interface is selected.
- Also it could be made configurable, which M2A relations of the child should be part of the tree. For example we have a O2M relation from parent to child. But the child contains two O2M relations. There could be a checkbox for each. And if you check one, and this one contains a O2M relation inside, there could be checkboxes again, this way you would have some sort of collapsed tree, that would expand as you choose, but it would only check the next nesting level after you enabled its parent. :)

I think my last point would supersede everything I wrote before and be the most clean and also safe way to go.


@rijkvanzanten No I don't think so. The nesting would be able to go as deep as there are O2M relations within the children. The only difference would be, that the tree would consist of different types of collections so drag and drop would be somewhat limited, as you can only drag children to parents of the correct collection, while on the current recursive collection approach everything can be a parent or a child of anything. But this behaviour would not go away, as you can still have rekursive collections aswell. This proposal is only about increasing the functionality so that simple O2M relations can be shown aswell and not only recursive ones

**Reply by LasseRosenow** [reply]
To quote myself:
> Also it could be made configurable, which M2A relations of the child should be part of the tree. For example we have a O2M relation from parent to child. But the child contains two O2M relations. There could be a checkbox for each. And if you check one, and this one contains a O2M relation inside, there could be checkboxes again, this way you would have some sort of collapsed tree, that would expand as you choose, but it would only check the next nesting level after you enabled its parent. :)

Here are some graphics on how I think this could work:

## Kontext
I create a O2M relation called "children".
The M contains some O2M relation itself: "content", "users" and "children".
So now you can choose which you would like to include in your tree.

## General Interface
![image](https://user-images.githubusercontent.com/10547444/129755629-0a2f19d8-108e-4e1c-acc9-d9d9e7547fdf.png)

## Different states
<table>
<tr><th>Only "content" is enabled, but content does not have any O2M relations itself</th>
<th>Now "children" is enabled aswell and will be part of the tree, but "children" also has some O2M relations itself: "users" and "children", but those are not enabled, so they wont be part of the tree</th></tr>
<tr>
<td>

![image](https://user-images.githubusercontent.com/10547444/129756361-8a0f13b6-0bdb-41b0-aa21-dcd89ae26449.png)

</td>
<td>

![image](https://user-images.githubusercontent.com/10547444/129757024-c4988a48-5cf0-4c98-9507-607e0c2cbc8f.png)

</td>
</tr>
</table>

Also I would like to add, that if a nested child is from the same collection, as its parent and it is enabled aka. a recursive one, I would say the right way would be not to expand it and show all its childrens, as this could go on for ever, but to apply the same rules as for its parent. This way this configuration does not limit recursive relations.

**Comment by lukas-schaetzle** [comment]
This is one of the few big features I'd like to see implemented before using directus in production. I guess a lot of people who'd like to see #3054 (tree view as layout) implemented would also favour this one here as it would also allow easy editing of the item relationships.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
